File: frontend/src/lib/api/remote-control.ts
import { api } from './client';
import type { ChannelKind } from '$lib/config/channels';
import type {
	ChannelCall,
	ChannelCatalogEntry,
	ChannelChat,
	ChannelChatUpdate,
	ChannelCommand,
	ChannelSettingsUpdate,
	ChannelStatus,
	ChannelVerifyResult,
	PairingApprove,
	PairingRequest
} from '$lib/types/remote-control';

const base = (channel: ChannelKind) => `/remote-control/channels/${channel}`;

export const remoteControlApi = {
	channels(): Promise<ChannelCatalogEntry[]> {
		return api.get<ChannelCatalogEntry[]>('/remote-control/channels');
	},

	status(channel: ChannelKind): Promise<ChannelStatus> {
		return api.get<ChannelStatus>(`${base(channel)}/status`);
	},

	update(channel: ChannelKind, body: ChannelSettingsUpdate): Promise<ChannelStatus> {
		return api.patch<ChannelStatus>(`${base(channel)}/settings`, body);
	},

	verify(channel: ChannelKind): Promise<ChannelVerifyResult> {
		return api.post<ChannelVerifyResult>(`${base(channel)}/verify`);
	},

	pending(channel: ChannelKind): Promise<PairingRequest[]> {
		return api.get<PairingRequest[]>(`${base(channel)}/pending`);
	},

	approve(channel: ChannelKind, code: string, body: PairingApprove): Promise<ChannelChat> {
		return api.post<ChannelChat>(
			`${base(channel)}/pending/${encodeURIComponent(code)}/approve`,
			body
		);
	},

	block(channel: ChannelKind, code: string): Promise<void> {
		return api.post<void>(`${base(channel)}/pending/${encodeURIComponent(code)}/block`);
	},

	chats(channel: ChannelKind): Promise<ChannelChat[]> {
		return api.get<ChannelChat[]>(`${base(channel)}/chats`);
	},

	updateChat(channel: ChannelKind, id: string, body: ChannelChatUpdate): Promise<ChannelChat> {
		return api.patch<ChannelChat>(`${base(channel)}/chats/${id}`, body);
	},

	revokeChat(channel: ChannelKind, id: string): Promise<ChannelChat> {
		return api.post<ChannelChat>(`${base(channel)}/chats/${id}/revoke`);
	},

	deleteChat(channel: ChannelKind, id: string): Promise<void> {
		return api.delete<void>(`${base(channel)}/chats/${id}`);
	},

	commands(channel: ChannelKind): Promise<ChannelCommand[]> {
		return api.get<ChannelCommand[]>(`${base(channel)}/commands`);
	},

	calls(channel: ChannelKind, limit = 200): Promise<ChannelCall[]> {
		return api.get<ChannelCall[]>(`${base(channel)}/calls?limit=${limit}`);
	}
};

File: api/app/api/v1/remote_control.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentSuperuser, CurrentUser
from app.core.database import get_session
from channels.models import (
    ChannelCatalogEntry,
    ChannelChatRead,
    ChannelChatUpdate,
    ChannelCommandRead,
    ChannelSettingsUpdate,
    ChannelStatus,
    ChannelVerifyResult,
    PairingApprove,
    PairingRequestRead,
)
from channels.service import (
    ChannelConfigError,
    ChannelNotFoundError,
    ChannelService,
    catalog,
)
from mcp.models import McpCallRead

router = APIRouter(prefix="/remote-control", tags=["remote-control"])

Session = Annotated[AsyncSession, Depends(get_session)]


def _service(session: AsyncSession, channel: str) -> ChannelService:
    try:
        return ChannelService(session, channel)
    except ChannelNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


def _guard(exc: ChannelConfigError) -> HTTPException:
    return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))


@router.get("/channels", response_model=list[ChannelCatalogEntry])
async def list_channels(_current_user: CurrentUser, session: Session):
    return await catalog(session)


@router.get("/channels/{channel}/status", response_model=ChannelStatus)
async def channel_status(_current_user: CurrentUser, session: Session, channel: str):
    return await _service(session, channel).status()


@router.patch("/channels/{channel}/settings", response_model=ChannelStatus)
async def update_channel(
    admin: CurrentSuperuser,
    session: Session,
    channel: str,
    body: ChannelSettingsUpdate,
):
    try:
        return await _service(session, channel).update(body, admin.id)
    except ChannelConfigError as exc:
        raise _guard(exc) from exc


@router.post("/channels/{channel}/verify", response_model=ChannelVerifyResult)
async def verify_channel(_admin: CurrentSuperuser, session: Session, channel: str):
    return await _service(session, channel).verify()


@router.get("/channels/{channel}/pending", response_model=list[PairingRequestRead])
async def pending_pairings(_admin: CurrentSuperuser, session: Session, channel: str):
    return await _service(session, channel).pending()


@router.post(
    "/channels/{channel}/pending/{code}/approve",
    response_model=ChannelChatRead,
    status_code=status.HTTP_201_CREATED,
)
async def approve_pairing(
    admin: CurrentSuperuser,
    session: Session,
    channel: str,
    code: str,
    body: PairingApprove,
):
    try:
        return await _service(session, channel).approve(code, body, admin.id)
    except ChannelConfigError as exc:
        raise _guard(exc) from exc


@router.post(
    "/channels/{channel}/pending/{code}/block", status_code=status.HTTP_204_NO_CONTENT
)
async def block_pairing(
    admin: CurrentSuperuser, session: Session, channel: str, code: str
):
    try:
        await _service(session, channel).block(code, admin.id)
    except ChannelConfigError as exc:
        raise _guard(exc) from exc


@router.get("/channels/{channel}/chats", response_model=list[ChannelChatRead])
async def list_chats(_admin: CurrentSuperuser, session: Session, channel: str):
    return await _service(session, channel).chats()


@router.patch("/channels/{channel}/chats/{chat_id}", response_model=ChannelChatRead)
async def update_chat(
    _admin: CurrentSuperuser,
    session: Session,
    channel: str,
    chat_id: UUID,
    body: ChannelChatUpdate,
):
    try:
        return await _service(session, channel).update_chat(chat_id, body)
    except ChannelConfigError as exc:
        raise _guard(exc) from exc


@router.post(
    "/channels/{channel}/chats/{chat_id}/revoke", response_model=ChannelChatRead
)
async def revoke_chat(
    _admin: CurrentSuperuser, session: Session, channel: str, chat_id: UUID
):
    try:
        return await _service(session, channel).revoke_chat(chat_id)
    except ChannelConfigError as exc:
        raise _guard(exc) from exc


@router.delete(
    "/channels/{channel}/chats/{chat_id}", status_code=status.HTTP_204_NO_CONTENT
)
async def delete_chat(
    _admin: CurrentSuperuser, session: Session, channel: str, chat_id: UUID
):
    try:
        await _service(session, channel).delete_chat(chat_id)
    except ChannelConfigError as exc:
        raise _guard(exc) from exc


@router.get("/channels/{channel}/commands", response_model=list[ChannelCommandRead])
async def list_commands(_current_user: CurrentUser, session: Session, channel: str):
    return _service(session, channel).commands()


@router.get("/channels/{channel}/calls", response_model=list[McpCallRead])
async def list_calls(
    _current_user: CurrentUser,
    session: Session,
    channel: str,
    limit: Annotated[int, Query(ge=1, le=200)] = 200,
):
    return await _service(session, channel).calls(limit)
